src/devices/machine/fdc37c665gt.h

```diff
@@ -1,37 +1,133 @@
 // license:BSD-3-Clause
-// copyright-holders:smf
-/*
-* fdc37c665gt.h
-*
-*/
+// copyright-holders: Samuele Zannoli, windyfairy
+/***************************************************************************
+
+FDC37C665GT.h
+
+SMSC FDC37C665GT High Performance Multi-Mode Parallel Port Super I/O Floppy Disk Controllers
+
+***************************************************************************/
 
 #ifndef MAME_MACHINE_FDC37C665GT_H
 #define MAME_MACHINE_FDC37C665GT_H
 
 #pragma once
 
-#include "ins8250.h"
+// floppy disk controller
+#include "machine/upd765.h"
+#include "imagedev/floppy.h"
+#include "formats/pc_dsk.h"
+// parallel port
+#include "machine/pc_lpt.h"
+// serial port
+#include "machine/ins8250.h"
 
 class fdc37c665gt_device : public device_t
 {
 public:
-	// construction/destruction
-	fdc37c665gt_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+	fdc37c665gt_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+		: fdc37c665gt_device(mconfig, tag, owner, clock, upd765_family_device::mode_t::AT)
+	{ }
+
+	fdc37c665gt_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock, upd765_family_device::mode_t floppy_mode);
+
+	// optional information overrides
+	virtual void device_add_mconfig(machine_config &config) override;
 
+	// to access io ports
 	uint8_t read(offs_t offset);
 	void write(offs_t offset, uint8_t data);
 
+	auto fintr() { return m_fintr_callback.bind(); }
+	auto fdrq() { return m_fdrq_callback.bind(); }
+	auto txd1() { return m_txd1_callback.bind(); }
+	auto ndtr1() { return m_ndtr1_callback.bind(); }
+	auto nrts1() { return m_nrts1_callback.bind(); }
+	auto txd2() { return m_txd2_callback.bind(); }
+	auto ndtr2() { return m_ndtr2_callback.bind(); }
+	auto nrts2() { return m_nrts2_callback.bind(); }
+
+	// for the internal floppy controller
+	DECLARE_WRITE_LINE_MEMBER(irq_floppy_w);
+
+	// for the internal parallel port
+	DECLARE_WRITE_LINE_MEMBER(irq_parallel_w);
+
+	// for the internal uarts
+	DECLARE_WRITE_LINE_MEMBER(irq_serial1_w);
+	DECLARE_WRITE_LINE_MEMBER(txd_serial1_w);
+	DECLARE_WRITE_LINE_MEMBER(dtr_serial1_w);
+	DECLARE_WRITE_LINE_MEMBER(rts_serial1_w);
+	DECLARE_WRITE_LINE_MEMBER(irq_serial2_w);
+	DECLARE_WRITE_LINE_MEMBER(txd_serial2_w);
+	DECLARE_WRITE_LINE_MEMBER(dtr_serial2_w);
+	DECLARE_WRITE_LINE_MEMBER(rts_serial2_w);
```

Comment:
Stuff that’s hooked up in your own machine config function doesn’t need to be public any more (since machine config functions became member functions).  It should be safe for these to be protected.